File: lib/java/opentoken/src/test/java/com/truveta/opentoken/io/parquet/PersonAttributesParquetReaderTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.io.parquet;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.HashMap;
import java.util.Map;
import java.util.NoSuchElementException;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
import org.apache.parquet.column.ParquetProperties.WriterVersion;
import org.apache.parquet.hadoop.ParquetFileWriter;
import org.apache.parquet.hadoop.example.ExampleParquetWriter;
import org.apache.parquet.hadoop.example.GroupWriteSupport;
import org.apache.parquet.schema.MessageTypeParser;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import com.truveta.opentoken.attributes.Attribute;
import com.truveta.opentoken.attributes.general.RecordIdAttribute;
import com.truveta.opentoken.attributes.person.FirstNameAttribute;
import com.truveta.opentoken.attributes.person.SocialSecurityNumberAttribute;

class PersonAttributesParquetReaderTest {

    private File tempFile;
    private String tempFilePath;

    @BeforeEach
    void setUp() throws IOException {
        tempFile = Files.createTempFile("test_data", ".parquet").toFile();
        tempFilePath = tempFile.getAbsolutePath();
    }

    @AfterEach
    void tearDown() {
        if (tempFile.exists()) {
            tempFile.delete();
        }
    }

    @Test
    void testReadParquet() throws Exception {
        try (PersonAttributesParquetWriter writer = new PersonAttributesParquetWriter(tempFilePath)) {

            Map<String, String> record1 = new HashMap<>();
            record1.put("RecordId", "1");
            record1.put("SocialSecurityNumber", "123-45-6789");
            record1.put("FirstName", "John");
            writer.writeAttributes(record1);

            Map<String, String> record2 = new HashMap<>();
            record2.put("RecordId", "2");
            record2.put("SocialSecurityNumber", "987-65-4321");
            record2.put("FirstName", "Jane");
            writer.writeAttributes(record2);
        }

        try (PersonAttributesParquetReader reader = new PersonAttributesParquetReader(tempFilePath)) {
            assertTrue(reader.hasNext());

            Map<Class<? extends Attribute>, String> firstRecord = reader.next();
            assertEquals("1", firstRecord.get(RecordIdAttribute.class));
            assertEquals("123-45-6789", firstRecord.get(SocialSecurityNumberAttribute.class));
            assertEquals("John", firstRecord.get(FirstNameAttribute.class));

            assertTrue(reader.hasNext());

            Map<Class<? extends Attribute>, String> secondRecord = reader.next();
            assertEquals("2", secondRecord.get(RecordIdAttribute.class));
            assertEquals("987-65-4321", secondRecord.get(SocialSecurityNumberAttribute.class));
            assertEquals("Jane", secondRecord.get(FirstNameAttribute.class));

            assertFalse(reader.hasNext());
        }
    }

    @Test
    void testReadEmptyParquet() throws Exception {
        var conf = new Configuration();
        String schemaString = "message PersonAttributes {\n" +
                "  required binary BirthDate (UTF8);\n" +
                "  required binary Gender (UTF8);\n" +
                "  required binary FirstName (UTF8);\n" +
                "  required binary SocialSecurityNumber (UTF8);\n" +
                "  required binary RecordId (UTF8);\n" +
                "  required binary PostalCode (UTF8);\n" +
                "  required binary LastName (UTF8);\n" +
                "}";
        var schema = MessageTypeParser.parseMessageType(schemaString);
        GroupWriteSupport.setSchema(schema, conf);
        Path path = new Path(tempFilePath);

        var writer = ExampleParquetWriter.builder(path)
                .withWriteMode(ParquetFileWriter.Mode.OVERWRITE)
                .withWriterVersion(WriterVersion.PARQUET_2_0)
                .withConf(conf)
                .withType(schema)
                .build();
        writer.close();

        try (PersonAttributesParquetReader reader = new PersonAttributesParquetReader(tempFilePath)) {
            assertFalse(reader.hasNext());
        }
    }

    @Test
    void testHasNext() throws Exception {
        try (PersonAttributesParquetWriter writer = new PersonAttributesParquetWriter(tempFilePath)) {

            Map<String, String> record1 = new HashMap<>();
            record1.put("RecordId", "1");
            record1.put("SocialSecurityNumber", "123-45-6789");
            record1.put("FirstName", "John");
            writer.writeAttributes(record1);
        }

        try (PersonAttributesParquetReader reader = new PersonAttributesParquetReader(tempFilePath)) {
            assertTrue(reader.hasNext());
            reader.next();
            assertFalse(reader.hasNext());
        }
    }

    @Test
    void testNext() throws Exception {
        try (PersonAttributesParquetWriter writer = new PersonAttributesParquetWriter(tempFilePath)) {

            Map<String, String> record1 = new HashMap<>();
            record1.put("RecordId", "1");
            record1.put("SocialSecurityNumber", "123-45-6789");
            record1.put("FirstName", "John");
            writer.writeAttributes(record1);
        }

        try (PersonAttributesParquetReader reader = new PersonAttributesParquetReader(tempFilePath)) {
            boolean hasNext = reader.hasNext();
            assertTrue(hasNext);

            Map<Class<? extends Attribute>, String> record = reader.next();
            assertNotNull(record);
            assertEquals("1", record.get(RecordIdAttribute.class));
            assertEquals("123-45-6789", record.get(SocialSecurityNumberAttribute.class));
            assertEquals("John", record.get(FirstNameAttribute.class));
        }
    }

    @Test
    void testClose() throws Exception {
        try (PersonAttributesParquetWriter writer = new PersonAttributesParquetWriter(tempFilePath)) {

            Map<String, String> record1 = new HashMap<>();
            record1.put("RecordId", "1");
            record1.put("SocialSecurityNumber", "123-45-6789");
            record1.put("FirstName", "John Doe");
            writer.writeAttributes(record1);
        }

        PersonAttributesParquetReader reader = new PersonAttributesParquetReader(tempFilePath);
        reader.close();
        assertThrows(NoSuchElementException.class, reader::hasNext);
        assertThrows(NoSuchElementException.class, reader::next);
    }

    @Test
    void testConstructorThrowsIOException() {
        String invalidtempFilePath = "non_existent_file.parquet";
        assertThrows(IOException.class, () -> new PersonAttributesParquetReader(invalidtempFilePath));
    }
}

File: lib/python/opentoken/src/test/opentoken/io/parquet/person_attributes_parquet_reader_test.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import os
import tempfile

import pytest
try:
    import pyarrow as pa
    import pyarrow.parquet as pq
except ImportError:
    pytest.skip("PyArrow required for Parquet tests", allow_module_level=True)

from opentoken.attributes.general.record_id_attribute import RecordIdAttribute
from opentoken.attributes.person.first_name_attribute import FirstNameAttribute
from opentoken.attributes.person.social_security_number_attribute import SocialSecurityNumberAttribute
from opentoken.io.parquet.person_attributes_parquet_reader import PersonAttributesParquetReader
from opentoken.io.parquet.person_attributes_parquet_writer import PersonAttributesParquetWriter


class TestPersonAttributesParquetReader:
    """Test cases for PersonAttributesParquetReader."""

    def setup_method(self):
        """Set up test fixtures before each test method."""
        self.temp_file = tempfile.NamedTemporaryFile(suffix='.parquet', delete=False)
        self.temp_file_path = self.temp_file.name
        self.temp_file.close()

    def teardown_method(self):
        """Clean up after each test method."""
        if os.path.exists(self.temp_file_path):
            os.unlink(self.temp_file_path)

    def test_read_parquet(self):
        """Test reading a Parquet file."""
        with PersonAttributesParquetWriter(self.temp_file_path) as writer:
            record1 = {
                "RecordId": "1",
                "SocialSecurityNumber": "123-45-6789",
                "FirstName": "John"
            }
            writer.write_attributes(record1)

            record2 = {
                "RecordId": "2",
                "SocialSecurityNumber": "987-65-4321",
                "FirstName": "Jane"
            }
            writer.write_attributes(record2)

        with PersonAttributesParquetReader(self.temp_file_path) as reader:
            # Test first record
            first_record = next(reader)
            assert first_record[RecordIdAttribute] == "1"
            assert first_record[SocialSecurityNumberAttribute] == "123-45-6789"
            assert first_record[FirstNameAttribute] == "John"

            # Test second record
            second_record = next(reader)
            assert second_record[RecordIdAttribute] == "2"
            assert second_record[SocialSecurityNumberAttribute] == "987-65-4321"
            assert second_record[FirstNameAttribute] == "Jane"

            # Test no more records
            with pytest.raises(StopIteration):
                next(reader)

    def test_read_empty_parquet(self):
        """Test reading an empty Parquet file."""
        # Create an empty Parquet file
        schema = pa.schema([
            ('BirthDate', pa.string()),
            ('Gender', pa.string()),
            ('FirstName', pa.string()),
            ('SocialSecurityNumber', pa.string()),
            ('RecordId', pa.string()),
            ('PostalCode', pa.string()),
            ('LastName', pa.string())
        ])
        
        empty_arrays = [pa.array([], type=field.type) for field in schema]
        table = pa.table(empty_arrays, schema=schema)
        pq.write_table(table, self.temp_file_path)

        with PersonAttributesParquetReader(self.temp_file_path) as reader:
            with pytest.raises(StopIteration):
                next(reader)

    def test_iterator_protocol(self):
        """Test iterator protocol."""
        with PersonAttributesParquetWriter(self.temp_file_path) as writer:
            record1 = {
                "RecordId": "1",
                "SocialSecurityNumber": "123-45-6789",
                "FirstName": "John"
            }
            writer.write_attributes(record1)

        with PersonAttributesParquetReader(self.temp_file_path) as reader:
            # Test that we can iterate
            for record in reader:
                assert record[RecordIdAttribute] == "1"
                assert record[SocialSecurityNumberAttribute] == "123-45-6789"
                assert record[FirstNameAttribute] == "John"
                break

    def test_next(self):
        """Test next method."""
        with PersonAttributesParquetWriter(self.temp_file_path) as writer:
            record1 = {
                "RecordId": "1",
                "SocialSecurityNumber": "123-45-6789",
                "FirstName": "John"
            }
            writer.write_attributes(record1)

        with PersonAttributesParquetReader(self.temp_file_path) as reader:
            record = next(reader)
            assert record is not None
            assert record[RecordIdAttribute] == "1"
            assert record[SocialSecurityNumberAttribute] == "123-45-6789"
            assert record[FirstNameAttribute] == "John"

    def test_close(self):
        """Test close method."""
        with PersonAttributesParquetWriter(self.temp_file_path) as writer:
            record1 = {
                "RecordId": "1",
                "SocialSecurityNumber": "123-45-6789",
                "FirstName": "John Doe"
            }
            writer.write_attributes(record1)

        reader = PersonAttributesParquetReader(self.temp_file_path)
        reader.close()
        
        # After closing, next should raise StopIteration
        with pytest.raises(StopIteration):
            next(reader)

    def test_constructor_throws_io_exception(self):
        """Test constructor throws IOError for non-existent file."""
        invalid_file_path = "non_existent_file.parquet"
        with pytest.raises(IOError):
            PersonAttributesParquetReader(invalid_file_path)